torch/_inductor/codegen/cpp_wrapper_cpu.py

```diff
@@ -253,6 +254,123 @@ def get_input_cpp_type(input):
             return DTYPE_TO_CPP[dtype]
         return f"ArrayRefTensor<{DTYPE_TO_CPP[input.get_dtype()]}>"
 
+    def generate_input_output_runtime_checks(self):
+        dtype_to_name = {
+            torch.uint8: (0, "unsigned char"),
+            torch.int8: (1, "signed char"),
+            torch.int16: (2, "short int"),
+            torch.int32: (3, "int"),
+            torch.int64: (4, "long int"),
+            torch.float16: (5, "c10::Half"),
+            torch.float32: (6, "float"),
+            torch.float64: (7, "double"),
+            # the following dtype(s) are not supported by cpp_wrapper
+            # torch.uint16, torch.uint32, torch.uint64, torch.uint32
+            torch.complex32: (8, "c10::complex<c10::Half>"),
+            torch.complex64: (9, "c10::complex<float>"),
+            torch.complex128: (10, "c10::complex<double>"),
+            torch.bool: (11, "bool"),
+            torch.bfloat16: (15, "c10::BFloat16"),
+            torch.float8_e4m3fn: (23, "c10::Float8_e4m3fn"),
+            torch.float8_e5m2: (24, "c10::Float8_e5m2"),
+            torch.float8_e4m3fnuz: (25, "c10::Float8_e4m3fnuz"),
+            torch.float8_e5m2fnuz: (26, "c10::Float8_e5m2fnuz"),
+        }
+
+        # In debug_compile mode, we generate checks to ensure the dtype/shape/stride of each
+        # real input/output tensor match ones provided at compile time via sample
+        # input/output.
+        def gen_check(handle_kind, idx, name, tensor):
+            self.prefix.writeline(f"auto {name} = {handle_kind}[{idx}];")
+            self.codegen_tensor_dtype_var_decl(self.prefix, name)
+            expected_dtype, expected_dtype_name = dtype_to_name[tensor.dtype]
```

Comment:
You can use `aoti_torch_dtype_*` to get `expected_dtype`, instead of creating another dict store the mapping. For  `expect_dtype_name`, I think it makes more sense to read from https://github.com/pytorch/pytorch/blob/88ebdbc97c103271766203df6662240e95a09b42/torch/_inductor/codegen/cpp.py#L83

Reply:
> You can use aoti_torch_dtype_* to get expected_dtype, instead of creating another dict store the mapping.

Hmm, if my understanding is correct, I think you are suggesting that we construct something like ```aoti_torch_dtype_{tensor.dtype}``` based on the dtype of the tensor? Seems it doesn't work, because we want to query the runtime dtype form the real input tensor instead of the tensor from the compilation time. In other words, we don't know the dtype of the real input at the runtime, and this is actually the purpose of inserting runtime checks - we want to check this "unknown" dtype is the same as the one used at the compilation time.

> For expect_dtype_name, I think it makes more sense to read from DTYPE_TO_ATEN

Good suggestion. Fixed. Thanks!

Reply:
Isn't `{name}_dtype` what you get as the real input dtype? I think expected_dtype is statically known? The reason I am suggesting that is because mapping name to a enum value can be fragile (although the dtype enum values are unlikely to change in reality).

Reply:
Oh, I see. You were suggesting we use ```aoti_torch_dtype_``` to retrieve the expected dtype. That's a good idea. Fixed. Thanks.